lambda_hook/tests/test_lambda.py

```diff
@@ -103,6 +102,28 @@ def test_ensure_bucket_unhandled_error(self):
         exc = cm.exception
         self.assertEqual(exc.response["Error"]["Code"], 500)
 
+    # This should fail, your task is to figure out why and
+    # make it pass.
+    def test_upload_lambda_functions(self):
+        # 1st call
+        self.stubber.add_response("head_bucket", {})
+        self.stubber.add_response("head_object", {})
+        self.stubber.add_response("put_object", {})
+
+        # 2nd call
+        self.stubber.add_response("head_bucket", {})
+        self.stubber.add_response("head_object", {
+            "ETag": '"b9b90449fe17ded2c9424367f0fd147e"' # correct hash for the files, you can trust this, including extra quotes
+        })
+        # should not call put_object again, so no stubbing
+
+        try:
+            with self.temp_directory_with_files() as tmp_dir:
+                with self.stubber:
+                    aws_lambda.upload_lambda_functions(self.s3, BUCKET_NAME, "things", tmp_dir.path)
+                    aws_lambda.upload_lambda_functions(self.s3, BUCKET_NAME, "things", tmp_dir.path)
+        finally:
+            tmp_dir.cleanup()
```

Comment:
what exactly causes this test to fail before the fix? (i don't know a lot about testing in python but i was expecting to see some kind of assertion-like thing)

Reply:
The test fails because the code is making a call to s3 with put_object when it shouldn't (because the files haven't changed). When you fix the code it doesn't make the call to s3 because hashes are the same and so it passes. But it would be better to explicitly put in the test assertMethodNotCalled(put_object) or something. I need to check how to that in python

Reply:
ohhh i see, so this test will fail bc it doesn't have that 2nd `put_object` stub?

Reply:
Yeeeh

Reply:
cool cool. i think the test would be more clear with something like a `assertMethodNotCalled` like you mentioned but i also think this works for now! non-blocking to me since the interviewer can always add more color there if needed